Add PySide6 to rosdep keys

samyarsadat opened the pull request:
Please add the following dependency to the rosdep database.

## Package name:

PySide6

## Package Upstream Source:

https://github.com/qtproject/pyside-pyside-setup

## Purpose of using this:

This package is a Python wrapper for Qt 6. It allows one to create Qt applications using Python.

## Links to Distribution Packages

<!-- Replace the REQUIRED areas with the URL to the package.  For IF AVAILABLE areas, either put in the URL to the package or state 'not available'.
More info at https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#guidelines-for-rosdep-rules -->

- PyPi: https://pypi.org/project/PySide6/

samyarsadat committed:
Added PySide6 with pip.

samyarsadat committed:
Removed incorrectly added quotation marks.

marcoag commented on rosdep/python.yaml:
```diff
@@ -8442,6 +8442,10 @@ python3-pyside2.qtwidgets:
     '8': null
   ubuntu: [python3-pyside2.qtwidgets]
+python3-pyside6-pip:
+  '*':
+    pip:
+      packages: [PySide6]
```
It seems like PySide6 is coming for next [Ubuntu](https://packages.ubuntu.com/search?keywords=pyside6&searchon=names&suite=all&section=all) and [Debian](https://packages.debian.org/search?keywords=pyside6&searchon=names&suite=all&section=all) distros. Since we rather not have pip packages when the system package are available, I would suggest to scope this to specific distros (feel free to add more):

```suggestion
  debian:
    bullseye:
      pip:
        packages: [PySide6]
    bookworm:
      pip:
        packages: [PySide6]
  ubuntu:
    focal:
      pip:
        packages: [PySide6]
    jammy:
      pip:
        packages: [PySide6]
    noble:
      pip:
        packages: [PySide6]
```

Feel free to add other distros.

samyarsadat replied:
I have incorporated your suggestion, and also added keys for RHEL 8, RHEL 9, and OSX.

marcoag commented:
Standard pip disclaimer: ROS packages that depend on pip keys cannot be released into a ROS distribution. They can only be depended on by from-source builds. Because of this, system packages are highly preferred to pip packages.

samyarsadat committed:
Scoped PySide6 to specific OSes and versions.

samyarsadat committed:
Fixed alphabetical order.

clalancette reviewed (commented):
This one ends up being tricky.

We know that system packages for pyside6 are coming; they are already available for [Ubuntu Oracular](https://packages.ubuntu.com/source/oracular/pyside6) and [Debian Sid](https://packages.debian.org/source/sid/pyside6).

However, as a matter of course we don't usually add in keys for pre-release distributions.

There are 2 ways we could go here:
1.  Take in this `python3-pyside6-pip` key for now, and later add in a `python3-pyside6` key.
2.  Right now go for a `python3-pyside6` key, and fill it out with `pip` for everything except for very new Ubuntu and Debian distributions.

Personally, I think we should go for 2, but I can be convinced otherwise.

@mjcarroll @alsora Any thoughts here?

samyarsadat commented:
I also think number 2 is a better option, but there is one small issue with it.
As with PySide2, the PySide6 binary packages are split up into different modules (ex. `python3-pyside6.qtqml`, `python3-pyside6.qtgui`, etc.), so keys for all of those would have to be added as well (at least for the latest Debian and Ubuntu distros that are going to use system packages).

clalancette commented:
> As with PySide2, the PySide6 binary packages are split up into different modules (ex. `python3-pyside6.qtqml`, `python3-pyside6.qtgui`, etc.), so keys for all of those would have to be added as well (at least for the latest Debian and Ubuntu distros that are going to use system packages).

Yeah, agreed.  However, we can do that piecemeal.  What particular parts of PySide6 are you interested in?

alsora commented:
I agree with option 2.
It also seems slightly more consistent with the rest of the file, where the `-pip` suffix is used only for those dependencies that always come from `pip`, while no suffix is used for dependencies where we use a mixture of `pip` and system packages.

samyarsadat commented:
> > As with PySide2, the PySide6 binary packages are split up into different modules (ex. `python3-pyside6.qtqml`, `python3-pyside6.qtgui`, etc.), so keys for all of those would have to be added as well (at least for the latest Debian and Ubuntu distros that are going to use system packages).
> 
> Yeah, agreed. However, we can do that piecemeal. What particular parts of PySide6 are you interested in?

That works. I currrently need `qtcore`, `qtgui`, `qtconcurrent`, `qtqml`, and `qtwidgets`.

Should there be one key named `python3-pyside6` that installs PySide6 thourgh pip on distros with no system packages and `libpyside6-dev`, `python3-pyside6.qtcore`, `python3-pyside6.qtgui`, `python3-pyside6.qtconcurrent`, `python3-pyside6.qtqml`, and `python3-pyside6.qtwidgets` on distros with system packages, or should there be individual keys for all of the different modules?

clalancette commented:
> or should there be individual keys for all of the different modules?

I would start with individual keys.  That gives packagers maximum control.  If it turns out that there is really a need for a "consolidated" key, we can always add that in later.

samyarsadat committed:
Add PySide6 system packages for Ubuntu Oracular & Debian Sid.

samyarsadat commented:
I've updated the `python3-pyside6` key and added all the new keys for the different modules.
I previously said that I needed `qtcore`, `qtgui`, `qtconcurrent`, `qtqml`, and `qtwidgets`, but I've added a few more:
`qtasyncio`, `qtcharts`, `qtdatavisualization`, `qttest`.

nuclearsandwich commented on rosdep/python.yaml:
```diff
@@ -8442,6 +8442,81 @@ python3-pyside2.qtwidgets:
+    bullseye:
+      pip:
+        packages: [PySide6]
+    sid: [libpyside6-dev]
```
```suggestion
    trixie: [libpyside6-dev]
    sid: [libpyside6-dev]
```

Add the current Debian Testing distribution as well. I would even consider dropping the `sid` definition altogether in favor of just trixie but don't mind it being there.

It's worth making the change for each of the definitions below as well.

nuclearsandwich replied:
Done in https://packages.debian.org/testing/python3-pyside6.qtasyncio

sloretz replied:
Looks like this made CI unhappy

```
    ERR: list out of alphabetical order line 8526.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8551.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8557.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8563.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8569.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8575.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8581.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8587.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8593.  'sid' should come before 'trixie'
    ERR: list out of alphabetical order line 8599.  'sid' should come before 'trixie'
```

nuclearsandwich committed:
Add trixie definitions for pyside6.

Trixie is the current testing distribution of Debian.

* https://packages.debian.org/trixie/python3-pyside6.qtasyncio
* https://packages.debian.org/trixie/python3-pyside6.qtcharts
* https://packages.debian.org/trixie/python3-pyside6.qtconcurrent
* https://packages.debian.org/trixie/python3-pyside6.qtcore
* https://packages.debian.org/trixie/python3-pyside6.qtdatavisualization
* https://packages.debian.org/trixie/python3-pyside6.qtgui
* https://packages.debian.org/trixie/python3-pyside6.qtqml
* https://packages.debian.org/trixie/python3-pyside6.qttest
* https://packages.debian.org/trixie/python3-pyside6.qtwidgets

nuclearsandwich commented:
@sloretz I modified the PR to include trixie definitions, would you please give it an updated review with my changes.

nuclearsandwich committed:
Change sort order.

I'd intentionally tried to keep these in release order rather than
lexicographical order but I forgot that the checker would dislike this.

clalancette committed:
Switch around the cases here.

In particular, make this forward-compatible by using
a star for the newer distributions, and either pip or
null for the older ones.  This way it will work going
forward.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Add exception for Fedora 39.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Alphabetize.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Another fix.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
I updated this to be more forwards-compatible.  With this in place, I'm happy, so I'm going to approve.  I would still like another approval before merging.

cottsay reviewed (approved):
Looks good. Could use an OS-level wildcard for the pip rule on `python3-pyside6` and drop the explicit osx and rhel rules if you like.

clalancette commented:
> Looks good. Could use an OS-level wildcard for the pip rule on `python3-pyside6` and drop the explicit osx and rhel rules if you like.

Thanks for the review.  I'm going to leave things as-is for now, we can always follow-up later on.
